src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionHandler.cs

```diff
@@ -0,0 +1,143 @@
+            HttpConnectionKey key = new HttpConnectionKey(request.RequestUri);
+            HttpConnectionPool pool;
+            if (_connectionPoolTable.TryGetValue(key, out pool))
+            {
+                connection = pool.GetConnection();
+            }
+
+            if (connection == null)
+            {
+                // No connection available in pool.  Create a new one.
+                connection = await CreateConnection(request, key, pool);
+            }
+
+            return await connection.SendAsync(request, cancellationToken);
+        }
+
+        private async Task<SslStream> EstablishSslConnection(string host, HttpRequestMessage request, Stream stream)
+        {
+            RemoteCertificateValidationCallback callback = null;
+            if (_serverCertificateCustomValidationCallback != null)
+            {
+                callback = (object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+                {
+                    return _serverCertificateCustomValidationCallback(request, certificate as X509Certificate2, chain, sslPolicyErrors);
+                };
+            }
+
+            SslStream sslStream = new SslStream(stream, false, callback);
+
+            try
+            {
+                // TODO: No cancellationToken?
+                await sslStream.AuthenticateAsClientAsync(host, _clientCertificates, _sslProtocols, _checkCertificateRevocationList);
+            }
+            catch (AuthenticationException ae)
+            {
+                // TODO: Tests expect HttpRequestException here.  Is that correct behavior?
+                sslStream.Dispose();
+                throw new HttpRequestException("could not establish SSL connection", ae);
+            }
+            catch (IOException ie)
+            {
+                // TODO: Tests expect HttpRequestException here.  Is that correct behavior?
+                sslStream.Dispose();
+                throw new HttpRequestException("could not establish SSL connection", ie);
+            }
+            catch (Exception)
+            {
+                sslStream.Dispose();
+                throw;
+            }
+
+            return sslStream;
+        }
+
+        private async Task<HttpConnection> CreateConnection(HttpRequestMessage request, HttpConnectionKey key, HttpConnectionPool pool)
+        {
+            Uri uri = request.RequestUri;
+
+            Stream stream = await ConnectHelper.ConnectAsync(uri.Host, uri.Port);
+
+            TransportContext transportContext = null;
+
+            if (uri.Scheme == "https")
+            {
+                SslStream sslStream = await EstablishSslConnection(uri.Host, request, stream);
+
+                stream = sslStream;
+                transportContext = sslStream.TransportContext;
+            }
+
+            if (pool == null)
+            {
+                pool = _connectionPoolTable.GetOrAdd(key, new HttpConnectionPool());
```

Comment:
Is the connection pool object expensive?  (Should that use the lambda instead of a direct ctor?)